ldpd/lde.c

```diff
@@ -1687,6 +1785,278 @@ lde_change_host_label(int af)
 		    NULL, 0);
 }
 
+void
+lde_change_advertise_to_filter(int af)
+{
+	struct lde_nbr  *ln;
+	struct fec      *f;
+	struct fec_node *fn;
+	char            *acl_name;
+	enum fec_type   fec_type;
+
+	/* advertise label for fecs to neighbors in this filter */
+	switch (af) {
+	case AF_INET:
+		acl_name = ldeconf->ipv4.acl_label_advertise_to;
+		fec_type = FEC_TYPE_IPV4;
+		break;
+	case AF_INET6:
+		acl_name = ldeconf->ipv6.acl_label_advertise_to;
+		fec_type = FEC_TYPE_IPV6;
+		break;
+	default:
+		fatalx("lde_change_advertise_to_filter: unknown af");
+	}
+
+	RB_FOREACH(ln, nbr_tree, &lde_nbrs) {
+		if (lde_acl_check(acl_name, af, (union ldpd_addr *)&ln->id, 32)
+		    != FILTER_PERMIT)
+			/* This neighbor is now filtered so send withdraw */
+			lde_send_labelwithdraw_wcard(ln, NO_LABEL);
+		else {
+			/* This neighbor is allowed in filter list, so
+			 * send labels
+			 */
+			RB_FOREACH(f, fec_tree, &ft) {
+				fn = (struct fec_node *)f;
+				if (fn->fec.type == fec_type &&
+				    fn->local_label != NO_LABEL)
+					lde_send_labelmapping(ln, fn, 0);
+			}
+			lde_imsg_compose_ldpe(IMSG_MAPPING_ADD_END,
+					      ln->peerid, 0, NULL, 0);
+		}
+	}
+}
+
+void
+lde_change_advertise_for_filter(int af)
+{
+	struct lde_nbr  *ln;
+	struct fec      *f;
+	struct fec_node *fn;
+	char            *acl_name;
+	union ldpd_addr *prefix;
+	uint8_t          plen;
+	struct lde_map  *me;
+
+	/* advertise labels for fecs matching this filter */
+	RB_FOREACH(ln, nbr_tree, &lde_nbrs) {
+		RB_FOREACH(f, fec_tree, &ft) {
+			fn = (struct fec_node *)f;
+			switch (af) {
+			case AF_INET:
+				if (fn->fec.type != FEC_TYPE_IPV4)
+					continue;
+				acl_name = ldeconf->ipv4.acl_label_advertise_for;
+				prefix = (union ldpd_addr *)
+					&fn->fec.u.ipv4.prefix;
+				plen = fn->fec.u.ipv4.prefixlen;
+				break;
+			case AF_INET6:
+				if (fn->fec.type != FEC_TYPE_IPV6)
+					continue;
+				acl_name = ldeconf->ipv6.acl_label_advertise_for;
+				prefix = (union ldpd_addr *)
+					&fn->fec.u.ipv6.prefix;
+				plen = fn->fec.u.ipv6.prefixlen;
+				break;
+			default:
+				fatalx("lde_change_egress_label: unknown af");
+			}
+
+			me = (struct lde_map *)fec_find(&ln->sent_map, &fn->fec);
+			if (me && lde_acl_check(acl_name, af, prefix, plen)
```

Comment:
The `acl_label_advertise_to` and `acl_label_advertise_for` filters are complementary. When either of them is updated, both need to be processed again for all FECs.

The same applies for the `acl_label_accept_for` and `acl_label_accept_from` filters.

Reply:
would this meet your reqirement
static void
lde_check_filter_af(int af, struct ldpd_af_conf *af_conf, 
    const char *filter_name)
{
	if (strcmp(af_conf->acl_label_allocate_for, filter_name) == 0)
		lde_change_allocate_filter(af);
	if ((strcmp(af_conf->acl_label_advertise_to, filter_name) == 0) 
	    || (strcmp(af_conf->acl_label_advertise_for, filter_name) == 0)) {
		lde_change_advertise_to_filter(af);
		lde_change_advertise_for_filter(af);
	}
	if ((strcmp(af_conf->acl_label_accept_for, filter_name) == 0)
	    || (strcmp(af_conf->acl_label_accept_from, filter_name) == 0)) { 
		lde_change_accept_for_filter(af);
		lde_change_accept_from_filter(af);
	}
	if (strcmp(af_conf->acl_label_expnull_for, filter_name) == 0)
		lde_change_expnull_for_filter(af);
}

Reply:
That wouldn't yield correct results either. For every FEC, both filters need to be checked in sequence (like done in `lde_check_mapping()`) before taking any action.

Reply:
I made the code changes you recommended but when I tried to test with both a advertise to and advertise for filter the code would not take both.   Only one filter is shown in show running config?

lynne-VirtualBox(config-ldp-af)# label local advertise for 4
lynne-VirtualBox(config-ldp-af)# label local advertise to 5


show running config:
mpls ldp
 router-id 2.2.2.2
 !
 address-family ipv4
  discovery transport-address 2.2.2.2
  label local advertise to 5      <-------- advertise from seems to be missing
  !
  interface r2-eth1
  !
  interface r2-eth2
  !
 exit-address-family
 !
!
I also added a temporary log to see what filters are actually being used and it seems that only the last one configured is used.    I guess that is why I didn't think they could be executed at the same time.   I did test with an allocate and advertise filter together and accept and advertise filters together. 

If you look at ldp_vty_label_advertise the code looks to me that when you set the advertise_to filter is will clear the advertise_from filter.    Please advise if I'm missing something here.

Reply:
You should use `label local advertise for 4 to 5` instead (everything in the same line).